<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Madeleine Price Ball -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

Issues with the Variant Centric Approach
========================================

This is a cautionary tale about the "variant call format" - known as
VCF - and how a variant-centric approach poses serious issues for
scientific usability of a genome data database. In this particular
example, we're examining the Harvard Personal Genome Project (PGP)
data as it is currently represented in Google Genomics' BigQuery
database (as of May 16 2014).

This story contains two vignettes based on real world data, an interlude,
a piece of speculative fiction, and a happy conclusion.

1. Frederick can't find participants with Factor V Leiden
2. Thomas confirms the amazing intelligence of PGP participants
3. Interlude: Casey updates the database
4. Emily is confused about variant frequencies
5. Ending: Casey retroactively makes data variant-agnostic
6. Appendix: Queries upon native CGI data

Frederick can't find participants with Factor V Leiden
------------------------------------------------------

Frederick is a new grad student excited by participatory research.
He sees that PGP Harvard participants are sharing both trait
information and genome data, and he's curious if there are
participants with genetic diseases they don't know about.

Fred's last rotation was in a lab that studied how blood clots,
so his first idea is to see if any participants carry
[Factor V Leiden](https://en.wikipedia.org/wiki/Factor_V_Leiden) -
and if so, whether they've reported the trait of
"Hereditary thrombophilia".

Factor V Leiden is caused by a G to A substitution in the protein
coding portion of the Factor V gene, a variation corresponding to
dbSNP ID rs6025. In build 37 this is located at chromosome 1,
position 169519049.

So Frederick runs the following on BigQuery:
```
SELECT
  call.gt,
  COUNT(call.gt) AS cnt,
FROM
  [google.com:biggene:pgp.variants]
WHERE
  ( contig_name = '1' AND
    start_pos = 169519049 )
GROUP BY
  call.gt
ORDER BY
  cnt DESC
```
And Frederick finds:

| call genotype | sample count | Frederick's notes               |
|---------------|--------------|---------------------------------|
| 1/1           | 167          | homozygous for variant genotype |
| 1/0           | 5            | heterozygous carrier of variant |

Frederick is baffled. "A 1 refers to the non-reference variant...
that means 167 participants are homozygous for Factor V Leiden?"

Dr. Brainy overhears and looks over Frederick's shoulder. "Oh,"
she says, "For this variant, the reference genome *has* Factor
V Leiden. So you're looking for people that are 0/0."

"Oooh." says Frederick. "That makes sense. It's pretty rare,
so I guess there aren't any homozygotes in the PGP."

"Well actually," says Dr. Brainy, "We can't tell."

"Why?" asks Frederick.

"Using tools provided by Complete Genomics -- the sequencer
of these genomes -- Google Genomics converted the data into
classic VCF format before loading it," responds Dr. Brainy.
"Unfortunately, the classic VCF format only reports
information where a participant has at least one variant.
So that's what got loaded into the database."

"Wait," responds Frederick. He quickly searches for more
information about the PGP data. "Look, there's 172 genomes
total. 167 and 5 add up to 172 - so we can tell there aren't
any homozygous reference participants."

"That's good. But if those numbers hadn't matched up, we
wouldn't really know if the remainder were homozygous. They
might simply be missing that data, whole genome data always
has some missing regions."

"Oh." Fred frowns. "So, like, we're never sure how many are
homozygous for reference? In any of the PGP data?!" he asks.
"Doesn't this make the data really weird ... maybe unusable...
if you're trying to analyze genotype/trait correlations?"

"I'm afraid so," concludes Dr. Brainy.

Thomas confirms amazing intelligence in the PGP cohort
------------------------------------------------------

Thomas is a precocious undergraduate in the lab who wants to
analyze genomes. He just read about [a variant predicted to
increase IQ by up to six points](http://www.economist.com/news/science-and-technology/21601809-potent-source-genetic-variation-cognitive-ability-has-just-been)!
"Wow," thinks Thomas, "If that's true... I wonder if it's
enriched in PGP participants. They have to pass this
complicated exam to be in the project, so they're probably
smarter than average."

The variation in this study corresponds to dbSNP ID rs9536314,
causing an amino acid substitution in the Klotho gene (KL F327V).
About 30% of people carry the variant. In build 37, this is an A
to G substition at chromosome 13, position 33628138.

So Thomas builds and runs the following on BigQuery:
```
SELECT
  call.gt,
  COUNT(call.gt) AS cnt
FROM
  [google.com:biggene:pgp.variants]
WHERE
  ( contig_name = '13' AND
    start_pos = 33628138 )
GROUP BY
  call.gt
ORDER BY
  cnt DESC
```

And Thomas finds:

| call genotype | sample count | Thomas's notes                          |
|---------------|--------------|-----------------------------------------|
| 1/0           | 32           | participants with 1 copy of the variant |
| 1/1           | 5            | participants with 2 copies of variant   |

"Holy crap!" exclaims Thomas. "Every single participant here has
at least one copy of this variant!"

"This really is the IQ gene, and the participants are hella-smart!"

Dr. Brainy overhears Thomas and smacks her head.

"What?" says Thomas.

Dr. Brainy sighs. "Go show that to Frederick."

Interlude: Casey updates the database
------------------------------------

Over at Google Genomics, Casey hears about the stories of
Thomas and Frederick. With a generous mug of coffee, Casey
gets to work.

Soon the data is updated: the genome imports now include a
step where each genome is examined at all known variant
positions. To do this, Casey creates a list combining all
variant positions in the batch, and those already known
about in the database. Then, matching-reference genotypes
are extracted for these positions (i.e. if the genome had
reference called) and the additional data is added to the
database.

"Problem solved," announces Casey.

Emily is confused about variant frequencies
-------------------------------------------

Emily is a seasoned grad student: she wants to defend,
and she's hoping to dig up some interesting observations
to publish through analysis of online data.

One question Emily thinks to ask: do variant frequencies vary
according to ancestry? For example, does Ashkenazi ancestry
mean fewer rare variants because of a
[bottleneck effect](https://en.wikipedia.org/wiki/Population_bottleneck)?
Using BigQuery to examine PGP and 1000 Genomes data, Emily
starts to plot histograms of variant frequencies in each
genome.

She soon notices a pattern, but it's bizarre. The PGP data is
an ongoing project and, for some reason, the more recent genomes
have fewer very rare variants? How could this be?

Emily suspects it's a data artifact, but she has no explanation.
She brings it to Dr. Brainy.

"Oh... yeah, I thought this might happen," mutters Dr. Brainy.

"Why? Do you know what's causing this phenomenon?" asks Emily.

"It's just that variants and genomes are hard," sighs Dr. Brainy.
"Casey fixed how they import genomes, but it's asymmetric. When
an old genome has a very rare variant, all the new genomes that get
added record that position as matching reference. But when the
new genome has a very rare variant, the opposite isn't true."

"I suppose," she continues, "the proper solution would be to run through
all the original data from the previous genomes to report reference
genotypes for any new variant positions added by the new genome."

"Or maybe actually having a database entry for every position in the
genome? Although how would you handle indels...." Her voice trails off.

"Hey," interjects Emily, "Isn't this like
[confirmation bias](https://en.wikipedia.org/wiki/Confirmation_bias)?
If we only record data for positions where some previous genome had
a variant, that means we'll tend to overestimate variant frequencies,
especially the rare ones? Because they're biased to come from smaller
sample sets."

"Yes," agrees Dr. Brainy, "That's a fair summary."

"This is hard," reflects Emily, "I feel bad for Casey!"

Ending: Casey retroactively makes data variant-agnostic
-------------------------------------------------------

Back at Google Genomics, Casey realizes a better solution for storing
the data. The original Complete Genomics data, and the gVCF format
created by Illumina, both report "matching reference" regions rather
than individual positions. Storing the *regions* in the database, and
reworking database queries to check for such fragments, would provide
the complete solution.

But going back to all the old data to redo data processing is tedious,
so Casey applies for one cycle of Retroactive Messaging time at the
Googleplex (a popular internal feature where you send messages to
your past self). The database is retroactively upgraded.

In Timeline 2, Emily examines the database and finds perfect data, an
interesting finding, and publishes it in PLoS Genetics!

Appendix: Queries upon native CGI data
-------------------------------------------------------------
At a later date, the native CGI data was also loaded into BigQuery.  For more detail, see the [provenance](../../provenance) of table [cgi_variants](https://bigquery.cloud.google.com/table/google.com:biggene:pgp.cgi_variants?pli=1).  Note that this data is for 174 individuals whereas table [variants](https://bigquery.cloud.google.com/table/google.com:biggene:pgp.variants?pli=1) only has data for 172 individuals.

```{r init, echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
require(bigrquery)
require(ggplot2)
require(dplyr)
require(xtable)
require(testthat)
billing_project <- "google.com:biggene" # put your projectID here
DisplayAndDispatchQuery <- function(queryUri) {
  sql <- readChar(queryUri, nchars=1e6)
  cat(sql)
  query_exec(project="google.com:biggene", dataset="1000genomes",
                    query=sql, billing=billing_project)  
}
```

First let's take a look at the data relevant to Klotho:
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/cgi_variants/klotho.sql")
```
Number of rows returned by this query: `r nrow(result)`.  We have one row for every indivudual in this dataset.

Examing the first few rows, we see:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result, n=10)), type="html", include.rownames=F)
```

And summarizing by genotype:
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/issues-with-the-variant-centric-approach/klotho-summary.sql")
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
We see that 135 individuals are homozygous reference but two individuals were no-calls at this genomic position.

Next let's look at the data relevant to Factor V Leiden:
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/issues-with-the-variant-centric-approach/factor-v-leiden.sql")
```
Number of rows returned by this query: `r nrow(result)`.

Examing the first few rows, we see:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result, n=10)), type="html", include.rownames=F)
```

And summarizing by genotype:
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/issues-with-the-variant-centric-approach/factor-v-leiden-summary.sql")
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
We see that no individuals are homozygous reference and no one responded _true_ for the _Hereditary thrombophilia includes Factor V Leiden and Prothrombin G20210A_ trait.