<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

Understanding Alternate Alleles in 1,000 Genomes VCF Data
========================================================

We know from the [FAQ](http://www.1000genomes.org/faq/are-all-genotype-calls-current-release-vcf-files-bi-allelic) that the 1,000 Genomes VCF data is [bi-allelic](http://www.1000genomes.org/faq/are-all-genotype-calls-current-release-vcf-files-bi-allelic) → meaning that each row in the source VCF has only one value in the ALT field.  So for each sample in a row, the genotype was called as either the reference or the single ALT value.  At any particular position in the genome we can have much more variation than a single alternate, so we need to understand how that is encoded in this data set.

```{r init, echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
require(bigrquery)
require(ggplot2)
require(dplyr)
require(xtable)
require(testthat)
project <- "google.com:biggene" # put your projectID here
DisplayAndDispatchQuery <- function(queryUri) {
  sql <- readChar(queryUri, nchars=1e6)
  cat(sql)
  query_exec(sql, project)
}
```

Let’s explore the question _“Is (contig, position, reference_bases) a unique key in the 1,000 Genomes Data?”_

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/chrom-pos-ref-dups.sql")
```
Number of rows returned by this query: `r nrow(result)`.

We see the first six tabular results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```
So we see from the data that the answer to our question is “No”.

So how many rows might we see per (contig, position, reference_bases) tuple?
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/count-chrom-pos-ref.sql")
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
So we see that for any particular (contig, position, reference_bases) tuple the vast majority have a single alternate allele and a few have two.

Let’s examine a few of the tuples with two alternate alleles more closely.
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/three-chrom-pos-ref-dups.sql")
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
From this small sample, it appears that the alternate allele is either a SNP or an INDEL.

Is that the case for all the records corresponding to duplicate (contig, position, reference_bases) tuples?
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/count-by-var-type-chrom-pos-ref-dups.sql")
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
It appears that for all records for duplicate (contig, position, reference_bases) tuples that we have a SNP and also an INDEL or SV.

For records corresponding to a unique (contig, position, reference_bases) tuple, are the variants always SNPs?
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/count-by-var-type-chrom-pos-ref-singles.sql")
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
And we see that the answer to our question is “No” - for records corresponding to a unique (contig, position, reference_bases) tuple, the variants are mostly SNPs but also INDELs and SVs.

So what does this all mean for a particular duplicate (contig, position, reference_bases) tuple for a particular sample at a particular genomic position?
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/sample-chrom-pos-ref-dups.sql")
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
We can see that HG00101 was called the same in both records but HG00100 was called differently.  So which is the [correct interpretation](http://vcftools.sourceforge.net/VCF-poster.pdf) for each allele at position 48515943 on chromosome 17?
```
first allele
xxxTxxxx

second allele
xxxGxxxx
or
xxxTGxxx
```
Let’s examine the quality, some INFO fields, and the genotype likelihoods a little more closely.
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/sample-likelihood.sql")
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
The [likelihoods](http://faculty.washington.edu/browning/beagle/intro-to-vcf.html) correspond to the REF/REF, REF/ALT, and ALT/ALT genotypes in that order.  See the [schema](https://bigquery.cloud.google.com/table/google.com:biggene:1000genomes.variants1kG?pli=1) for details about the other fields.

So a question for our users who have much experience in this domain, which variant is more likely for the second allele of HG00100?

### But we digress . . .

Our original question was _“Is (contig, position, reference_bases) a unique key in the 1,000 Genomes Data?”_ which we know is false.  So which columns do constitute a unique key?

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/not-quite-unique-key.sql")
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
Not quite.  We see a few structural variant deletions called at the same position.

Let's add in the `end` column:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/unique-key.sql")
```

```{r warning=FALSE, comment=NA}
print(expect_true(is.null(result)))
```

And now we have it, a unique key is: (contig, position, reference_bases, alternate_bases, vt, end)

Lastly, what is a minimal unique key?
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/understanding-alternate-alleles/minimal-unique-key.sql")
```

```{r warning=FALSE, comment=NA}
print(expect_true(is.null(result)))
```

We see that a minimal unique key is: (contig, position, alternate_bases, end) or alternatively (contig, position, end, vt)
